sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,117 @@
+# Known Scalability Issues with Kubernetes Services
+
+_by Shyam JVS, Google Inc (with inputs from Brian Grant & Wojciech Tyczynski)_
+
+**June 2018**
+
+This document serves as a catalog of issues we've known/discovered with kubernetes services as of today, focusing on their scalability/performance. The purpose of the document is to make the information common knowledge for the community, so we can work together towards improving it. Listing them below in no particular order.
+
+
+## Performance of iptables when large number of service ports exist
+
+#### Issue
+
+Iptables can be slow in the presence of a large number of services. As the number of service ports increases, the KUBE-SERVICES chain gets longer and since it’s also evaluated very frequently that can bring down the performance. Also, it seems that manipulating a chain with lots of rules (e.g. KUBE-SERVICES) takes noticeably longer than others.
+
+#### Relevant issues/links
+
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+
+#### Possible Solution(s)
+
+- Moving from iptables to IPVS may help here if that works stably. We have the IPVS alternative [implemented](https://github.com/kubernetes/kubernetes/pull/46580), but it still hasn’t gone to GA. Also we need to measure/confirm performance benefits from it.
+
+
+## Kube-proxy slow in updating iptables when large number of backends exist
+
+#### Issue
+
+When we’re running services worth a total of 120k endpoints (backends) on our 5k-node scalability tests, we’re noticing that kube-proxy is timing out while trying to do iptables-restore due to failing to acquire lock over iptables.
```

Comment:
We're running 150k endpoints now.

Reply:
Yes, but if we're failing at a smaller no. we should mention that IMO.

Reply:
We don't currently test at that limit, but we have evidence from past that we see this issue at that number too (https://github.com/kubernetes/kubernetes/issues/48107). It may have changed now (but I'm not sure if some fixes were made for this bug in the meanwhile). So I feel we should have the lower number here to be rigorous.

Reply:
FWIW in terms of "timing out while trying to acquire lock", there's a second problem: `iptables-restore` grabs the lock *before* it parses its input, and simply parsing tens of thousands of iptables rules takes a noticeable amount of time. So if two iptables commands start at the same time, the first might grab the lock, and then start parsing its input, and burn through half of the other iptables command's `--wait` time before it even gets to the point of passing the rules off to the kernel.

This is not easy to fix because of the way that `iptables-restore` was written to share code with the main `iptables` binary, and no one is working on fixing it because the official plan is to move to nftables instead. If the world ends up moving to iptables-over-eBPF rather than nftables then this will probably need to be fixed at some point.

Reply:
I'm fine with having more rigorous number here. What I'm asking about is - why 120k?

If I would be reading it and would say 120k, I guess I would silently assume that 100k is fine (otherwise you would probably give more round number). Do we know that? (I think we don't).

So maybe it's safer to write 100k (if we are going this way?)

Reply:
@danwinship I've incorporated those very useful leads. Thanks and PTAL.

> What I'm asking about is - why 120k?

@wojtek-t Because it's a known limit for which we have evidence that we seen the issue. Anyway, I've now changed it to `> 100k` - lgty?